DSCResources/MSFT_xRobocopy/MSFT_xRobocopy.psm1

```diff
@@ -157,25 +140,41 @@ function Test-TargetResource
     #  4: mismatched files/directories
     #  8: copy errors and retries exceeded
     # 16: serious error
-    if ($result -ge 0 -and $result -lt 8) {[system.boolean]$result = $true}
-    else {[system.boolean]$result = $false}
+    if ($result -ge 1 -and $result -lt 8) 
+    {
+        Write-Verbose "Source and destination are out of sync"
+        [system.boolean]$result = $false
+    }
+    elseif ($result -eq 0)
+    {
+        Write-Verbose "Source and destination are completely synchronized"
+        [system.boolean]$result = $true
+    }
+    else
+    {
+        throw "robocopy test returned with errors, exit code: $result!"
+    }
     
     $result
 }
 
 # Helper Functions
 
 function Invoke-RobocopyTest {
-param (
-    [parameter(Mandatory = $true)]
-    [System.String]
-    $source,
+    param (
+        [parameter(Mandatory = $true)]
+        [System.String]
+        $source,
 
-    [parameter(Mandatory = $true)]
-    [System.String]
-    $destination
-)
-    $output = & robocopy.exe /L $source $destination
+        [parameter(Mandatory = $true)]
+        [System.String]
+        $destination,
+
+        [System.String]
+        $Arguments
+    )
+
+    $output = Invoke-Expression "&robocopy.exe /L `"$Source`" `"$Destination`" $Arguments"

```

Comment:
Hi,
I agree that Invoke-Expresion is not best, I have copied it from Set function.Wanted to get the bug fix quickly, without reimplemented the resource.
In my fork I have another branch were I'm working on better implementation of the resource that would conform to PowerShell best practices.
To be able to use argument array, we need to change AditionalArgs parameter to be array of strings.
I'm almost finished with new version, I'll submit new version in couple of days for review.
Thanks.


Sent from my Samsung Galaxy smartphone.<div>
</div><div>
</div><!-- originalMessage --><div>-------- Original message --------</div><div>From: Norberto Arrieta <notifications@github.com> </div><div>Date: 30/04/2016  01:05  (GMT+02:00) </div><div>To: PowerShell/xRobocopy <xRobocopy@noreply.github.com> </div><div>Cc: Arturas Kuznecovas <arturas.k@hotmail.com>, Mention <mention@noreply.github.com> </div><div>Subject: Re: [PowerShell/xRobocopy] Dev ak (#16) </div><div>
</div>
>
> -    [parameter(Mandatory = $true)]
> -    [System.String]
> -    $destination
> -)
> -    $output = & robocopy.exe /L $source $destination
> +        [parameter(Mandatory = $true)]
> +        [System.String]
> +        $destination,
> +
> +        [System.String]
> +        $Arguments
> +    )
> +
> +    $output = Invoke-Expression "&robocopy.exe /L `"$Source`" `"$Destination`" $Arguments"

Use of Invoke-Expression is discouraged, since it might open the possibility of script injection. I think in this case you can avoid Invoke-Expresion, &, and quotes if you instead use an array for the arguments and then use splatting (@), e.g.

PS> $source = 'path with space'
PS> $target = 'pathwithoutspace'
PS> $arguments = @('/L', '/V')
PS> robocopy.exe $source $target @arguments
        ...
    Source : D:\tmp\path with space\
    Dest : D:\tmp\pathwithoutspace\
    ...
    Options : *.* /V /L /DCOPY:DA /COPY:DAT /R:1000000 /W:30
    ...


---
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub:
https://github.com/PowerShell/xRobocopy/pull/16/files/bb423cf248ff1a44ef98c47389604961602f3f78#r61651086